File: outco/frequencyCount/unique.js
/*
 * Utilize the frequency count pattern to solve these problems.  Use a Hash Set
 * or an Array instead of a Hash Table where applicable.
 *
 * Unique
 *
 * Given an array of integers, return an array with all duplicates removed.*
 *
 * Parameters
 * Input: arr {Array of Integers}
 * Output: {Array of Integers}
 *
 * Constraints:
 *
 * Time: O(N)
 * Space: O(N)
 *
 * Examples
 * [1, 2, 4, 4, 5, 6] --> [1, 2, 4, 5, 6]
 * [1, 1, 2, 2, 3, 3]' --> [1, 2, 3]
 * [1, 2, 3, 1, 2] --> [1, 2, 3]
 */

'use strict';

// init array
// Only in cases where an array does not include an element
// append the element to array
// return array
// Time Complexity: O(N)
// Space Complexity: O(N)
// Auxiliary Complexity: O(N + 1) 
// Space Complexity == Auxiliary Complexity + Input Space Complexity

// Successfully tested the solution on 12/10/22.

function unique(arr) {
  let array = [];
  for (let i = 0; i < arr.length; i++){
    if (!array.includes(arr[i])){
      array.push(arr[i]);
    }
  }
  return array;
}

console.log(unique([1,1,2,2,3,3,4])) // [1,2,3,4]

File: c++journey/outco_in_cpp/frequencyCount/unique.cpp
/**
 *
 * Unique
 *
 * Given an array/vector of integers, return an array/vector with all duplicates removed.*
 *
 * Parameters
 * Input: arr {array/vector of Integers}
 * Output: {array/vector of Integers}
 *
 * Constraints
 *
 * Time: O(N)
 * Space: O(N)
 *
 * Examples
 * [1, 2, 4, 4, 5, 6] --> [1, 2, 4, 5, 6]
 * [1, 1, 2, 2, 3, 3]' --> [1, 2, 3]
 * [1, 2, 3, 1, 2] --> [1, 2, 3]
 */

#include <array>
#include <cmath>
#include <iterator>
#include <string>
#include <vector>
#include <algorithm>
#include <unordered_set>
#include <unordered_map>
#include <iostream>
#include <array>
using namespace std;

vector<int> unique(vector<int> vec){
    unordered_set<int> uniquesSet;
    for (int valueInt : vec){
        uniquesSet.insert(valueInt);
    }
    vector<int> uniquesVec;
    uniquesVec.reserve(uniquesSet.size());
    copy(uniquesSet.begin(), uniquesSet.end(), back_inserter(uniquesVec));
    return uniquesVec;
}   

void test(){
    vector<int> resultVec1 = unique({1, 1, 2, 2, 3, 3, 4, 4}); // expect 1,2,3,4,5
    cout << "Result vector 1: " << endl;
    for (int valueInt : resultVec1){
        cout << valueInt << " ";
    }
    vector<int> resultVec2 = unique({1, 2, 3, 1, 2}); // expect 1,2,3,
    cout << "\nResult vector 2: " << endl;
    for (int valueInt : resultVec2){
        cout << valueInt << " ";
    }
}

int main(){
    test();
    return 0;
}